namerd/iface/destination/src/main/scala/io/buoyant/namerd/iface/destination/DestinationService.scala

```diff
@@ -0,0 +1,196 @@
+      case Some(p) => Buf.Utf8(p.substring(0, p.indexOf(".")))
+    }
+
+    val name = Path(Seq(Buf.Utf8(pathPfx), path): _*)
+    val events = interpreter.bind(Dtab.empty, name).map(_.eval).flatMap {
+      case None => Activity.value(Var.value(Addr.Neg))
+      case Some(boundSet) =>
+        val set = Var.collect(boundSet.map(_.addr).toList).map(foldAddr)
+        Activity.value(set)
+    }.run.flatMap {
+      case Activity.Failed(e) => Var(Addr.Failed(e))
+      case Activity.Pending => Var(Addr.Pending)
+      case Activity.Ok(t) => t
+    }.changes.collect {
+      case Addr.Failed(e) => Throw(e)
+      case Addr.Neg => Throw(new Throwable(s"endpoints for service ${name.show} doesn't exist"))
+      case Addr.Bound(addresses, _) => Return(addresses)
+    }
+
+    val updateEvents = eventFlatMap[Diff[TrySet, Address], Ev[Update]](events.diff, {
+      case FailedDiff(e) =>
+        log.debug(s"address set encountered FailedDiff: $e")
+        Seq(Val(mkNoEndpointsUpdate(false)))
+      case SetDiff(add, remove) =>
+        val addUpdate = if (add.nonEmpty) Seq(Val(mkAddUpdate(add))) else Nil
+        val removeUpdate = if (remove.nonEmpty) Seq(Val(mkRemoveUpdate(remove))) else Nil
+        addUpdate ++ removeUpdate
+    })
+    val totalUpdates = prependEventOnCondition[Ev[Update]](updateEvents, Val(mkNoEndpointsUpdate(true)),
+      {
+        case Val(Update(Some(Add(_)))) => true
+        case _ => false
+      })
+    VarEventStream(totalUpdates)
+  }
+
+  private[this] def foldAddr(addrs: List[Addr]): Addr = {
+    addrs.fold(Addr.Pending) {
+      // nonPending take priority
+      case (Addr.Pending, nonPending) => nonPending
+      case (nonPending, Addr.Pending) => nonPending
+      // Bound addresses combined
+      case (Addr.Bound(a, _), Addr.Bound(b, _)) => Addr.Bound(a ++ b)
+      case (a@Addr.Bound(_, _), _) => a
+      case (_, a@Addr.Bound(_, _)) => a
+      // otherwise, just use the first state (e.g. in a list of negs or fails)
+      case (a, _) => a
+    }
+  }
+}
+
+object DestinationService {
+  private[this] def inetAddressToInt(ip: InetAddress): Int = {
+    ip.getAddress.reverse.zipWithIndex.foldRight(0) { (b, acc) =>
+      val pow = b._2 * 8
+      val bitShift = 1 << pow
+      acc + (b._1.toInt * bitShift)
+    }
+  }
+
+  //creates a sequence of events from one event based on function f
+  private[destination] def eventFlatMap[T, U](ev: Event[T], f: T => Seq[U]): Event[U] = new Event[U] {
+    def register(w: Witness[U]): Closable = {
+      ev.respond { t =>
+        for (u <- f(t)) w.notify(u)
+      }
+    }
+  }
+
+  //Prepends an init event to an already existing event only once.
+  private[destination] def prependEventOnCondition[T](ev: Event[T], init: T, f: T => Boolean): Event[T] = new Event[T] {
+    val prependOnce: AtomicBoolean = new AtomicBoolean(false)
+    def register(w: Witness[T]): Closable = {
+      ev.respond { t =>
+        if (f(t) && !prependOnce.get) {
```

Comment:
this can allow multiple threads in before the AB is set to true.  Consider something like:

```
// if this is already set, we can skip ahead
if (!prependOnce.get) { 
  // check f(t) first and if it returns true then do CAS
  if (f(t) && prependOnce.compareAndSet(false, true)) {
    w.notify(init)
  }
}
w.notify(t)
```

Reply:
Thanks, I think I ran into this while running my tests on k8s but couldn't figure out what was causing it. This change fixed it.